CoreAlgorithmPlugins/src/au/gov/asd/tac/constellation/plugins/algorithms/sna/similarity/CosineSimilarityPlugin.java

```diff
@@ -109,94 +109,128 @@ public void edit(final GraphWriteMethods graph, final PluginInteraction interact
-
-                    if (intersection.cardinality() < minCommonFeatures) {
-                        continue;
-                    }
-
-                    final int vertexOneId = graph.getVertex(vertexOnePosition);
-                    final int[] neighboursOne = neighbourWeights[vertexOnePosition];
-
-                    final int vertexTwoId = graph.getVertex(vertexTwoPosition);
-                    final int[] neighboursTwo = neighbourWeights[vertexTwoPosition];
-
-                    final float neighbourDotProduct = dot(neighboursOne, neighboursTwo);
-                    final float neighboursMagnitude = magnitude(neighboursOne) * magnitude(neighboursTwo);
-                    final float cosineSimilarity = neighboursMagnitude == 0 ? 0 : neighbourDotProduct / neighboursMagnitude;
-                    cosineSimilarities.put(Tuple.create(vertexOneId, vertexTwoId), cosineSimilarity);
                 }
             }
         }
-
-        // update the graph with cosine similarity values
-        SimilarityUtilities.addScoresToGraph(graph, cosineSimilarities, COSINE_SIMILARITY_ATTRIBUTE);
-
         // complete with schema
         PluginExecution.withPlugin(VisualSchemaPluginRegistry.COMPLETE_SCHEMA).executeNow(graph);
     }
-
-    private float dot(final int[] vectorOne, final int[] vectorTwo) {
-        assert vectorOne.length == vectorTwo.length;
-
-        float dot = 0;
-        for (int index = 0; index < vectorOne.length; index++) {
-            dot += vectorOne[index] * vectorTwo[index];
+    
+    private class VertexWithNeighbours {
+        private final int vertexId;
+        private final HashMap<Integer, Integer> neighbourWeightsMap;
+        private BitSet neighbours;
+        private final boolean selected;
+        private Float magnitude = null;
+
+        public VertexWithNeighbours(final int vertexId, final int vertexCount, final boolean selected) {
+            this.neighbourWeightsMap = new HashMap<>();
+            this.vertexId = vertexId;
+            this.selected = selected;
+            this.neighbours = new BitSet(vertexCount);
         }
-
-        return dot;
+           
+        private void addNeighbour(final int neighbourPosition, final int additionalWeight) {
+            int currentWeight = neighbourWeightsMap.getOrDefault(neighbourPosition, 0);
+            neighbourWeightsMap.put(neighbourPosition,currentWeight + additionalWeight);
+            neighbours.set(neighbourPosition, true);
+        }
+        
+        private float getMagnitude() {
+            if (magnitude == null) {
+                magnitude = calculateMagnitude();
+            }
+            return magnitude;
+        }
+        
+        private float calculateMagnitude() {
+            float mag = 0;
+            for(int neighbourWeight : neighbourWeightsMap.values()) {
+                mag += Math.pow(neighbourWeight, 2);
+            }
+            return (float) Math.sqrt(mag);
+        }
```

Comment:
From what I can see of its usage, I don't think `magnitude` needs to be a field. Potentially instead, calls of `getMagnitude()`, can be replaced with a call of `calculateMagnitude()` (thus omitting the need for `getMagnitude()` as well). If you move the call to the right place then you should only need to call the function once (which from what I can tell is the reason it has been done the way it has)

Reply:
@Nova-2119 If I recall, you initially instantiated `magnitude` to null so as far as I could tell, the null check in `getMagnitude()` was purely to run `calculateMagnitude()` the first time and so didn't appear to be needed for subsequent calls. Subsequent calls on `getMagnitude()` would still do that null check though so my idea really about moving that `calculateMagnitude()' out of the function so that a null check wasn't required everytime.

Obviously you've made changes since so it doesn't really apply now but those were my thoughts at the time